### PR TITLE
Select reading mode automatically. Deprecate READ_MODE and SET_FS_CONFIG options

### PR DESCRIPTION
**Pull request by michazag**

### CODE REVIEW

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala [hunk 1]
```diff
@@ -343,78 +343,76 @@ object KustoDataSourceUtils {
...
+        Some(KustoStorageParameters(storageAccount.get, storageSecret.get, storageContainer.get, storageSecretIsAccountKey))
       }
     }
   }
 
-  private[kusto] def getAndValidateTransientStorageParameters(storageAccount: Option[String],
-                                                              storageContainer: Option[String],
-                                                              storageAccountSecret: Option[String],
-                                                              storageSecretIsAccountKey: Boolean): KustoStorageParameters = {
+  private[kusto] def tryGetAndValidateTransientStorageParameters(storageAccount: Option[String],
```
Comment (ohadbitt) [inline connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala]: i would revert the renaming
usually in kusto a method with try prefix i surrounding the implementation with try catch- but here we throw

File: connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala [hunk 2]
```diff
@@ -40,34 +39,55 @@ private[kusto] case class KustoRelation(kustoCoordinates: KustoCoordinates,
...
+    if (count == 0)  {
+      sparkSession.emptyDataFrame.rdd
+    }
+    else {
+      val useLeanMode = KDSU.isLeanReadMode(count, storageParameters.isDefined, readOptions.forcedReadMode)
+      if (useLeanMode) {
+        KustoReader.leanBuildScan(
+          kustoClient,
+          KustoReadRequest(sparkSession, schema, kustoCoordinates, query, authentication, timeout)
+        )
```
[the remark this answers is not in the document]
Comment (michazag) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala]: No, the implementation is completely different.
If you mean that we could overload and have the same name that's true, but than the branching logic would just move one layer further. This is actually what top-level "buildscan" does

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoAzureFsSetupCache.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      // Entry exists but with a different secret - remove it and update
+      storageAccountKeyMap.remove(account)
+      secreteCached = ""
+    }
+
+    if (secreteCached.isEmpty || checkIfRefreshNeeded(utcNow)) {
+      storageAccountKeyMap += (account -> secret)
+      lastRefresh = utcNow
+
+      return false
```
[the remark this answers is not in the document]
Comment (michazag) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoAzureFsSetupCache.scala]: Actually, it is needed. 
I replaced below with "else true" - this way it is not needed

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoAzureFsSetupCache.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      // Entry exists but with a different secret - remove it and update
+      storageSasMap.remove(key)
+      secreteCached = ""
+    }
+
+    if (secreteCached.isEmpty || checkIfRefreshNeeded(utcNow)) {
+      storageSasMap += (key -> secret)
+      lastRefresh = utcNow
+
+      return false
```
[the remark this answers is not in the document]
Comment (michazag) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoAzureFsSetupCache.scala]: Same :)

File: connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala (hunk 1 above)
Comment (michazag) [reply connector/src/main/scala/com/microsoft/kusto/spark/utils/KustoDataSourceUtils.scala]: Changed to: 
getAndValidateTransientStorageParametersIfExist

**Pull request merged**